crates/evm/core/src/backend/mod.rs

```diff
@@ -1356,19 +1358,28 @@ impl DatabaseExt for Backend {
         mut env: Env,
         journaled_state: &mut JournaledState,
         inspector: &mut dyn InspectorExt,
+        inspect_ctx: Box<dyn Any>,
     ) -> eyre::Result<()> {
         trace!(?tx, "execute signed transaction");
 
         self.commit(journaled_state.state.clone());
 
         let res = {
             configure_tx_req_env(&mut env, tx, None)?;
-            let env = self.env_with_handler_cfg(env);
+            let mut env = self.env_with_handler_cfg(env);
 
             let mut db = self.clone();
-            let mut evm = new_evm_with_inspector(&mut db, env, inspector);
-            evm.context.evm.journaled_state.depth = journaled_state.depth + 1;
-            evm.transact()?
+            match db.strategy.runner.inspect(&mut db, &mut env, inspector, inspect_ctx) {
```

Comment:
This seems to be too complex. Not only are we splitting up the logic in disjoint places but the replacement code does not refer to the original code here. As per the strategy pattern we replace a single EVM call context via strategy call almost verbatim.
`inspect_ctx` shouldn't have been required to be set for EVM - it adds unnecessary indirection that diverges from upstream.

It's then preferable to abstract out the entire `transact_from_tx` as `zksync_transact_from_tx` and call that instead.